### PR TITLE
update pdfjs to latest 4.6.82

### PR DESCRIPTION
**Pull request by liudongmiao**
1. build our viewer.js with the source of pdf.js (it should be easy if we switch to gulp, use DefinePlugin to mock it too ugly)
2. use the global event bus, and remove _boundEvents (api change in pdfjs)

should close https://github.com/FirstTimeInForever/intellij-pdf-viewer/pull/102

### CODE REVIEW

**Comment by liudongmiao** [comment]
before update, there are different pdf.js versions:
- cmaps, pdf.worker.js, pdf_viewer.css, 2.6.347, via pdfjs-dist in package.json
- images, locale, 2.5.207, via pdf.js in postinstall.js
- viewer.css, 2.7.570, with little modification
- viewer.html (index.html), viewer.js, 2.8.335, with little modification
// viewer.html and viewer.js doesn't change much from 2.8.335 to 2.12.313

after update, the versions are updated to the same version.

File: web-view/bootstrap/index.html
Comment (liudongmiao) [inline web-view/bootstrap/index.html:1]: This file is synced with viewer.html in pdf.js: https://github.com/mozilla/pdf.js/blob/v4.6.82/web/viewer.html

File: web-view/bootstrap/package.json [hunk 1]
```diff
@@ -4,17 +4,19 @@
   "description": "",
   "scripts": {
     "postinstall": "node postinstall.js",
-    "build": "npx webpack --mode=development"
+    "build": "npx webpack --mode=production"
```
Comment (liudongmiao) [inline web-view/bootstrap/package.json:7]: the source in pdf.js use `PDFJSDev`, so we can only use production.

File: web-view/bootstrap/src/ViewerBootstrapper.ts [hunk 2]
```diff
@@ -1,28 +1,29 @@
-const {AppOptions} = require("pdfjs-dist/lib/web/app_options");
-const {PDFViewerApplication} = require("pdfjs-dist/lib/web/app");
-import {GenericExternalServices} from "./support/GenericExternalServices"
+// @ts-ignore
+import { AppOptions } from "pdf.js/web/app_options";
+Object.defineProperty(window, "PDFViewerApplicationOptions", {
```
Comment (liudongmiao) [inline web-view/bootstrap/src/ViewerBootstrapper.ts:3]: we need to expose PDFViewerApplicationOptions early for PDFJSDev, see below.

File: web-view/bootstrap/src/ViewerBootstrapper.ts [hunk 3]
```diff
@@ -1,28 +1,29 @@
...
-    });
     return PDFViewerApplication;
   }
 
   static load(fileUrl: String | null = null): Promise<void> {
     return new Promise(resolve => {
       AppOptions.set("defaultUrl", fileUrl);
       const config = getViewerConfiguration();
-      config.eventBus.on("documentloaded", () => resolve());
+      PDFViewerApplication.initializedPromise.then(function () {
```
Comment (liudongmiao) [inline web-view/bootstrap/src/ViewerBootstrapper.ts:24]: see https://github.com/mozilla/pdf.js/wiki/Third-party-viewer-usage#initialization-promise

File: web-view/bootstrap/src/support/ViewerConfiguration.ts
Comment (liudongmiao) [inline web-view/bootstrap/src/support/ViewerConfiguration.ts:1]: this file is synced with https://github.com/mozilla/pdf.js/blob/v4.6.82/web/viewer.js

File: web-view/bootstrap/src/viewer.css
Comment (liudongmiao) [inline web-view/bootstrap/src/viewer.css:1]: we use viewer.css from https://github.com/mozilla/pdf.js/blob/v4.6.82/web/viewer.css

File: web-view/bootstrap/webpack.config.js [hunk 4]
```diff
@@ -13,23 +14,99 @@ module.exports = {
     ]
   },
   resolve: {
+    alias: {
```
Comment (liudongmiao) [inline web-view/bootstrap/webpack.config.js:17]: see https://github.com/mozilla/pdf.js/blob/v4.6.82/gulpfile.mjs#L185

File: web-view/bootstrap/webpack.config.js [hunk 5]
```diff
@@ -13,23 +14,99 @@ module.exports = {
...
+          } else if (code === "COMPONENTS") {
+            return false;
+          }
+          throw new Error("PDFJSDev-test-" + code);
+        },
+        eval: function (code) {
+          if (code === "BUNDLE_VERSION") {
+            return "4.6.82";
+          } else if (code === "DEFAULT_PREFERENCES") {
+            return window.PDFViewerApplicationOptions.getAll(0x80, true);
```
Comment (liudongmiao) [inline web-view/bootstrap/webpack.config.js:88]: we don't copy the static content of `DEFAULT_PREFERENCES`, then should expose `PDFViewerApplicationOptions` very early.

File: web-view/bootstrap/webpack.config.js [hunk 6]
```diff
@@ -13,23 +14,99 @@ module.exports = {
...
+          } else if (code === "DEFAULT_FTL") {
+            return "";
```
Comment (liudongmiao) [inline web-view/bootstrap/webpack.config.js:90]: it's a fallback, so it's ok to return empty.

File: web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/Application.kt [hunk 7]
```diff
@@ -220,27 +220,6 @@ class Application(private val viewer: ViewerAdapter) {
   }
 
   private fun rebindStandardEvents() {
-    viewer.viewerApp.asDynamic().unbindWindowEvents()
-    val boundEvents = viewer.viewerApp.asDynamic()._boundEvents
```
Comment (liudongmiao) [inline web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/Application.kt:224]: _boundEvents is removed, see commit log 062acd1ef93892b326cdcb503e8b94bd40810d1f and https://github.com/mozilla/pdf.js/commit/702ee7b1e1775b6bfd57856beea77f650d6fa38a

**Review by liudongmiao — COMMENTED** [review]
@PHPirates Please help to review.

File: web-view/bootstrap/src/ViewerBootstrapper.ts (hunk 3 above)
Comment (liudongmiao) [reply web-view/bootstrap/src/ViewerBootstrapper.ts:24]: and https://github.com/mozilla/pdf.js/commit/08256e6795bbec885746cba1c198c0b4c5f84982

File: web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/Application.kt [hunk 8]
```diff
@@ -220,27 +220,6 @@ class Application(private val viewer: ViewerAdapter) {
...
-    boundEvents.windowBeforePrint = {
-      viewer.viewerApp.eventBus.dispatch("beforeprint", json("source" to window))
-    }
-    boundEvents.windowAfterPrint = {
-      viewer.viewerApp.eventBus.dispatch("afterprint", json("source" to window))
-    }
-    window.addEventListener("resize", boundEvents.windowResize.unsafeCast<EventListener>())
-    window.addEventListener("hashchange", boundEvents.windowHashChange.unsafeCast<EventListener>())
-    window.addEventListener("beforeprint", boundEvents.windowBeforePrint.unsafeCast<EventListener>())
-    window.addEventListener("afterprint", boundEvents.windowAfterPrint.unsafeCast<EventListener>())
```
Comment (liudongmiao) [inline web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/Application.kt:243]: we should think about this, use a better way to sync state.

Comment (PHPirates) [reply web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/Application.kt:243]: I agree this is not ideal, but what other way would you have in mind? We need to handle keyboard events in some way and link them to the viewer state.

File: web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/Application.kt (hunk 7 above)
Comment (PHPirates) [reply web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/Application.kt:224]: I see, I'm not quite sure why this code was here but everything seems to work fine still

**Review by PHPirates — APPROVED** [review]
Amazing, thanks a lot! Really appreciate the help with the javascript side of things, as I mainly have experience with writing intellij plugins in Kotlin.

File: web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/Application.kt (hunk 8 above)
Comment (liudongmiao) [reply web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/Application.kt:243]: The previous implementation for them is dispatch on event bus, and the default handler is event bus too.

When I scroll between page, the ViewState is changed, however, no subscriber of `PdfViewStateChangedListener.TOPIC` works (it's in PdfDocumentPageStatusBarWidget, however, no log at all).

**Comment by liudongmiao** [comment]
I'm thinking of remove [pdfjs-dist](https://www.npmjs.com/package/pdfjs-dist) in npm, and use the [pdf.js releases](https://github.com/mozilla/pdf.js/releases) .
pdfjs-dist in npm never distribute the default viewer, and has no plan.

https://github.com/mozilla/pdf.js/issues/5609
https://github.com/mozilla/pdf.js/issues/5709
https://github.com/mozilla/pdf.js/issues/6545
https://github.com/mozilla/pdf.js/issues/7203
https://github.com/mozilla/pdf.js/issues/7780
https://github.com/mozilla/pdf.js/issues/9127
https://github.com/mozilla/pdf.js/issues/9144
https://github.com/mozilla/pdf.js/issues/9210
https://github.com/mozilla/pdf.js/issues/16997

If we choose pdf.js releases, it would be simple in bootstrap.
we can only inject our viewer.js and fixes.css.

File: web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/Application.kt (hunk 8 above)
Comment (PHPirates) [reply web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/Application.kt:243]: Ah, the id of PdfDocumentPageStatusBarWidget was incorrect so the widget was broken, fixed that in #115. Well spotted, thanks!

**Comment by PHPirates** [comment]
Thanks for the links, strange decision they made! I agree intwould be a lot better if we could use the source including the default viewer directly, that would make updating a lot easier. A pull request would be greatly appreciated!

I'll merge this one already so at least we have a working version.

**Pull request merged**